internal/state/execution.go

```diff
@@ -416,43 +414,51 @@ func buildLastCommitInfo(block *types.Block, store Store, initialHeight int64) a
 	}
 }
 
-// extendedCommitInfo expects a CommitInfo struct along with all of the
-// original votes relating to that commit, including their vote extensions. The
-// order of votes does not matter.
-func extendedCommitInfo(c abci.CommitInfo, votes []*types.Vote) abci.ExtendedCommitInfo {
-	if len(c.Votes) != len(votes) {
-		panic(fmt.Sprintf("extendedCommitInfo: number of votes from commit differ from the number of votes supplied (%d != %d)", len(c.Votes), len(votes)))
-	}
-	votesByVal := make(map[string]*types.Vote)
-	for _, vote := range votes {
-		if vote != nil {
-			valAddr := vote.ValidatorAddress.String()
-			if _, ok := votesByVal[valAddr]; ok {
-				panic(fmt.Sprintf("extendedCommitInfo: found duplicate vote for validator with address %s", valAddr))
-			}
-			votesByVal[valAddr] = vote
-		}
+// buildLastExtendedCommitInfo expects an ExtendedCommitInfo, which includes
+// the original precommit votes along with their vote extensions from the last
+// commit. It populates the corresponding ABCI protobuf structures using the
+// validator set obtained from the store at the height of the last extended
+// commit.
```

Comment:
Optional readability nit (upshot: front-loading the part about what the function does before explanations)
```suggestion
// buildLastExtendedCommitInfo populates an ABCI extended commit from the
// corresponding Tendermint extended commit ec, using the stored validator set from
// ec.  It requires ec to include the  original precommit votes along
// with the vote extensions from the last commit.
//
// For heights below the initial height, for which we do not have the
// required data, it returns an empty record.
```

YMMV.